### PR TITLE
Add open source vs commercial

### PR DESCRIPTION
**Pull request by mooreds**
This is expert advice and a blog post.

### CODE REVIEW

File: site/learn/expert-advice/identity-basics/open-source-vs-commercial.md [hunk 1]
```diff
@@ -0,0 +1,127 @@
...
+
+The caveat to that price tag is that you generally also have to run this software on your own infrastructure, so there's an implicit cost there. However, if you're already running your own infrastructure, then in many cases, an open-source authentication provider can be added at no direct monetary cost to you.
+
+While this price tag is attractive, it's important not to forget that open-source providers come with added maintenance and security responsibility when compared to proprietary solutions, as we've already discussed.
+
+## Commercial Authentication Providers
+
+An alternative to open-source, commercial authentication providers (such as [FusionAuth](https://fusionauth.io/), [Okta](https://www.okta.com/), and [Auth0](https://auth0.com/)) are popular because they take care of all the ambiguity and self-reliance that come with open source.
+
+But just because commercial providers handle many of open source's problems doesn't mean there aren't tradeoffs that need to be considered with them as well.
```
Comment (jobannon) [inline site/learn/expert-advice/identity-basics/open-source-vs-commercial.md:76]: This sentence sounds a bit wordy.

File: site/learn/expert-advice/identity-basics/open-source-vs-commercial.md [hunk 2]
```diff
@@ -0,0 +1,127 @@
...
+
+This is another thing you don't have to worry about with a commercial provider, rather than being responsible for reviewing all these licenses if you went with an open-source solution.
+
+### Maintenance Responsibility
+
+This is one of the main benefits of a commercial provider. They are responsible for all the maintenance of their product. Because you're paying for the product directly, this is one area you no longer have to worry about. You are trading money for time.
+
+However, unlike with open source, you have less freedom to make fixes yourself. So if you have a feature or a fix that you believe should be in the product, the most you can do is try to convince someone at the vendor to get that prioritized. With an open-source solution, you could fix it yourself, if you are willing to accept the maintenance burden.
```
Comment (jobannon) [inline site/learn/expert-advice/identity-basics/open-source-vs-commercial.md:102]: I think that you could add a tidbit about an "SOW" if you truly needed a feature developed specific to your use case on the commercial side.

File: site/learn/expert-advice/identity-basics/open-source-vs-commercial.md [hunk 3]
```diff
@@ -0,0 +1,127 @@
...
+
+### Cost
+
+One of the more notable differences between open source and commercial solutions is cost. Because of the benefits detailed above, there is usually a cost to a commercial solution, whereas many open-source solutions are provided free of charge and maintained by the community. 
+
+However, some commercial software offers free tiers with limited support. For example, Auth0 allows you a certain number of users for free. FusionAuth has a community edition allowing you an unlimited number of users if you self-host.
+
+Depending on what stage your company is in, this cost can either be something you can absorb or a dealbreaker.
```
Comment (jobannon) [inline site/learn/expert-advice/identity-basics/open-source-vs-commercial.md:118]: **Perhaps (ad-libbing here):**

Another item to consider is the pricing structure of the commercial solution.  Some commercial products may prove cost-effective to large corporations, with high usage, but terribly improbable for a startup wanting a lowering buy-in point.  Or more commonly, commercial providers provide different cost tiers depending on usage which may simply be untenable to certain use-cases as a product scales.  What started as 0 dollars a month on Monday may cost 12K a month on Monday a year from now.

**Also perhaps:**
Do your homework.  A good deal is only a good deal if you can say "yes" to everything you are saying "no" to.  A lower cost of initial entry, but many hours logged switching providers as the corporation scales is no deal at all.  You get what you pay for - fair is fair.

**Review by jobannon — COMMENTED** [review]
Interesting article!  Stylistic observations/preferences, as usual 😄

File: site/_posts/2021-04-08-open-source-auth-considerations.md [hunk 4]
```diff
@@ -0,0 +1,67 @@
+---
+layout: blog-post
+title: Open source auth provider considerations
```
Comment (dpolstra) [inline site/_posts/2021-04-08-open-source-auth-considerations.md:3]: Maybe: the pitfalls of open source auth providers 
Or: the downsides of choosing open source auth providers

Comment (dpolstra) [reply site/_posts/2021-04-08-open-source-auth-considerations.md:3]: maybe we don't want to be _so_ negative, but I think it sets better expectations than 'considerations'

Comment (dpolstra) [reply site/_posts/2021-04-08-open-source-auth-considerations.md:3]: maybe: What to consider before choosing an open source auth provider

File: site/_posts/2021-04-08-open-source-auth-considerations.md [hunk 5]
```diff
@@ -0,0 +1,67 @@
...
+
+And second, open-source projects tend to "work in public," meaning discussions about release frequency, notes from planning meetings, and more may be publicly accessible. Depending on how the project is structured, you may even be able to be part of these meetings. Some projects have this information in their README file or some similar place. You'd rarely get this kind of transparency with a more proprietary provider. 
+
+However, it is worth noting the distinction between open source and open process. A project can be open source but take no feedback from the community; some of Apple's open source projects are like this. On the other hand, commercial software companies can develop “in the open” and even make the source available for inspection using source-available licenses, without allowing for the freedoms that are key to open source licensing. When you evaluate an open-source project, make sure you know what the community feedback process is, if any.
```
Comment (dpolstra) [inline site/_posts/2021-04-08-open-source-auth-considerations.md:33]: commercial software companies **_(like FusionAuth)_** can develop “in the open”

File: site/_posts/2021-04-08-open-source-auth-considerations.md [hunk 6]
```diff
@@ -0,0 +1,67 @@
...
+
+Some open-source projects, especially if they are backed by a company, support dual licensing. This means that you can choose one license or another as fits your business model. The first license may be the GPL or a similar restrictive license, while the other may be a commercial license with fewer restrictions that costs money. 
+
+Always review the license for any solution.
+
+### Maintenance Responsibility
+
+In a true open-source solution, the "community" is responsible for submitting fixes they want to see included in the project. Feature requests from the community tend to need to be contributed to by the community itself, meaning you as the customer bear more of the maintenance burden. The amount of testing done for a new release varies based on the effort the community puts forth.
```
Comment (dpolstra) [inline site/_posts/2021-04-08-open-source-auth-considerations.md:45]: tighten up the second sentence. feels odd

File: site/_posts/2021-04-08-open-source-auth-considerations.md [hunk 7]
```diff
@@ -0,0 +1,67 @@
...
+
+On the flip side, it could mean you have more say in the product roadmap if you're able to contribute fixes in a way that is acceptable to the community. Heck, you could even pay a developer to spend most of their day writing code for this system, which would buy you a lot of influence. Consider how much bandwidth your team has to contribute to this process; it might be a dealbreaker.
+
+You can also, if you need to, make a change in your own version of the project. You can then submit it to the maintainers. If it meets their quality standards and goal for the project, it can be integrated; this is also known as upstreaming. However, if it doesn't get incorporated, you are stuck supporting a critical feature and making sure that future releases don't break it.
```
Comment (dpolstra) [inline site/_posts/2021-04-08-open-source-auth-considerations.md:49]: ~if you need to~

File: site/_posts/2021-04-08-open-source-auth-considerations.md [hunk 8]
```diff
@@ -0,0 +1,67 @@
...
+However, there are also [instances of vulnerabilities remaining open for months](https://github.com/keycloak/keycloak/pull/7612). But if the bug is severe and an open-source project is well supported or popular, the timeline may shrink to days or weeks.
+
+In the end, you (and the maintainers) are ultimately responsible for the security of the software. Just like with maintenance responsibility, if you don't have time for your team to frequently review the security issues as they're reported and fixed and then upgrade your system, this may mean that open source isn't the best option for you.
+
+### Cost
+
+This is where many open-source solutions really shine. If you're running an open-source authentication provider, the cost for the provider itself in many cases is nothing. Zip. Nada. Who doesn't like free?
+
+The caveat to that price tag is that you generally also have to run this software on your own infrastructure, so there's an implicit cost there. However, if you're already running your own infrastructure, then in many cases, an open-source authentication provider can be added at no direct monetary cost to you.
+
```
Comment (dpolstra) [inline site/_posts/2021-04-08-open-source-auth-considerations.md:64]: Maybe add something around costing development time to getting it to work for your system?

File: site/learn/expert-advice/identity-basics/open-source-vs-commercial.md [hunk 9]
```diff
@@ -0,0 +1,127 @@
...
+author: Keanan Koppenhaver
+image: advice/due-diligence/authentication-as-a-service-due-diligence-tips-header-image.png
+category: Identity Basics
+date: 2021-04-15
+dateModified: 2021-04-15
+---
+
+When it comes to building any web or mobile application, authentication is one of those areas where you generally don't want to take on the risk yourself. Third-party authentication providers have made this their entire business and dedicate lots of time and money to make sure their solutions are performant, easy to use, and most of all, secure. However, some of these providers make the source code for their solutions available and others keep their source code proprietary.
+
+Choosing one of these types of authentication providers over the other is not always a cut-and-dried decision, as there are benefits to each. Support, release frequency, relicensing capability, maintenance, who is responsible for security, and cost are all factors you should consider when making this decision.
```
Comment (dpolstra) [inline site/learn/expert-advice/identity-basics/open-source-vs-commercial.md:14]: ~who is responsible for~ security

File: site/learn/expert-advice/identity-basics/open-source-vs-commercial.md [hunk 10]
```diff
@@ -0,0 +1,127 @@
...
+
+### Relicensing
+
+Commercial providers tend to have either more obscure or more restrictive licenses. You won't be able to take the code and do anything you want with it. If you are building a typical web application, this shouldn't be an issue, but if you are embedding or redistributing your code with the auth system, that may violate the license. Ask your vendor. 
+
+Many of the issues of relicensing that come with open-source technologies don't apply to commercial authentication providers, because they usually provide a single license that goes with their product. They handle the licensing of any component parts in-house and usually have legal teams or at the very least an attorney that's reviewing all of the different aspects of software licensing.
```
Comment (dpolstra) [inline site/learn/expert-advice/identity-basics/open-source-vs-commercial.md:94]: legal teams or ~at the very least~ an attorney

File: site/learn/expert-advice/identity-basics/open-source-vs-commercial.md (hunk 2 above)
Comment (dpolstra) [inline site/learn/expert-advice/identity-basics/open-source-vs-commercial.md:102]: We should add the community aspect of FusionAuth with prioritizing features. And I agree that we should add an SOW section too

File: site/learn/expert-advice/identity-basics/open-source-vs-commercial.md [hunk 11]
```diff
@@ -0,0 +1,127 @@
...
+
+This is another thing you don't have to worry about with a commercial provider, rather than being responsible for reviewing all these licenses if you went with an open-source solution.
+
+### Maintenance Responsibility
+
+This is one of the main benefits of a commercial provider. They are responsible for all the maintenance of their product. Because you're paying for the product directly, this is one area you no longer have to worry about. You are trading money for time.
+
+However, unlike with open source, you have less freedom to make fixes yourself. So if you have a feature or a fix that you believe should be in the product, the most you can do is try to convince someone at the vendor to get that prioritized. With an open-source solution, you could fix it yourself, if you are willing to accept the maintenance burden.
+
+If you are worried about access to the source code if the commercial vendor goes out of business, ask about a source code escrow. Some providers are willing to add a provision to their contract with you.
```
Comment (dpolstra) [inline site/learn/expert-advice/identity-basics/open-source-vs-commercial.md:104]: Some providers **_(like FusionAuth)_** are willing

File: site/learn/expert-advice/identity-basics/open-source-vs-commercial.md [hunk 12]
```diff
@@ -0,0 +1,127 @@
...
+
+### Security Responsibility
+
+As with maintenance responsibility, the responsibility for the security of a commercial product is on the vendor. They will be the ones applying patches and (hopefully) doing frequent security audits to ensure their product is secure. These could include running internal security reviews, a bug bounty program, or hiring a pentesting firm to test the product's security.
```
Comment (dpolstra) [inline site/learn/expert-advice/identity-basics/open-source-vs-commercial.md:108]: patches and ~(hopefully)~ doing frequent security audits

File: site/learn/expert-advice/identity-basics/open-source-vs-commercial.md (hunk 3 above)
Comment (dpolstra) [inline site/learn/expert-advice/identity-basics/open-source-vs-commercial.md:118]: Maybe add something about cost coming with ease of use and features that you wouldn't get with an open source solution

Comment (dpolstra) [reply site/learn/expert-advice/identity-basics/open-source-vs-commercial.md:118]: We could also elaborate on this in "The Final Showdown" section

File: site/_posts/2021-04-08-open-source-auth-considerations.md (hunk 4 above)
Comment (mooreds) [reply site/_posts/2021-04-08-open-source-auth-considerations.md:3]: `What to consider before choosing an open source auth provider` seems fair.

File: site/learn/expert-advice/identity-basics/open-source-vs-commercial.md (hunk 2 above)
Comment (mooreds) [reply site/learn/expert-advice/identity-basics/open-source-vs-commercial.md:102]: Great point!

File: site/learn/expert-advice/identity-basics/open-source-vs-commercial.md (hunk 3 above)
Comment (mooreds) [reply site/learn/expert-advice/identity-basics/open-source-vs-commercial.md:118]: Great points. I incorporated some of these.

**Pull request merged**